File: java/observer/NewsLetter/subject/Newsletter.java
package subject;

import observer.Observer;

import java.util.ArrayList;

public class Newsletter implements Subject {

    private String name;
    private ArrayList<Observer> observers;

    public Newsletter() { observers = new ArrayList<>(); }

    @Override
    public void registerObserver(Observer newObserver) {
        observers.add(newObserver);
    }

    @Override
    public void removeObserver(Observer removeObserver) {
        observers.remove(removeObserver);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer: observers) {
            observer.update(name);
        }
    }

    public void setName(String name) {
        this.name = name;
        notifyObservers();
    }

}

File: kotlin/observer/NewsLetter/subject/Newsletter.kt
package subject

import observer.Observer

class Newsletter : Subject {

    private var name: String? = null
    private var observers: ArrayList<Observer>? = null

    init {
        observers = ArrayList()
    }

    override fun registerObserver(newObserver: Observer) {
        observers!!.add(newObserver)
    }

    override fun removeObserver(removeObserver: Observer) {
        observers!!.remove(removeObserver)
    }

    override fun notifyObservers() {
        for (observer in observers!!) {
            observer.update(name!!)
        }
    }

    fun setName(name: String) {
        this.name = name
        notifyObservers()
    }
}